AppController/djinn.rb

```diff
@@ -783,56 +777,94 @@ def kill(stop_deployment, secret)
+      if node.class != Hash
+        msg = "Error: node structure is not a Hash."
+        Djinn.log_error(msg)
+        return msg
+      end
+      if !node['public_ip'] || !node['private_ip'] || !node['jobs'] ||
+        !node['instance_id']
+        msg = "Error: node layout is missing information #{node}."
+        Djinn.log_error(msg)
+        return msg
+      elsif node['public_ip'].empty? || node['private_ip'].empty? ||
+         node['jobs'].empty? || node['instance_id'].empty?
+        msg = "Error: node layout is missing information #{node}."
+        Djinn.log_error(msg)
+        return msg
+      end
+    }
 
-    if apps.class != Array
-      msg = "Error: apps wasn't an Array, but was a " + apps.class.to_s
+    # options is an array that we're converting to hash tables, so we need
+    # to make sure that every key maps to a value e.g., ['foo', 'bar']
+    # becomes {'foo' => 'bar'} so we need to make sure that the array has
+    # an even number of elements.
+    if options.class != String
+      msg = "Error: options wasn't a String, but was a " +
+        options.class.to_s
       Djinn.log_error(msg)
       return msg
     end
-
-    # credentials is an array that we're converting to
-    # hash tables, so we need to make sure that every key maps to a value
-    # e.g., ['foo', 'bar'] becomes {'foo' => 'bar'}
-    # so we need to make sure that the array has an even number of elements
-    if database_credentials.length.odd?
-      msg = "Error: DB Credentials wasn't of even length: Len = " + \
-        "#{database_credentials.length}"
+    begin
+      @options = JSON.load(options)
+    rescue JSON::ParserError => e
+      @options = nil
+      Djinn.log_debug("Got exception parsing JSON options.")
+    end
+    if @options.nil? || @options.empty? || @options.class != Hash
+      msg = "Error: @options is nil or not in the proper format."
       Djinn.log_error(msg)
       return msg
     end
 
-    possible_credentials = Hash[*database_credentials]
-    unless valid_format_for_credentials(possible_credentials)
-      return "Error: Credential format wrong"
-    end
-
-    keyname = possible_credentials["keyname"]
-    @options = possible_credentials
-    @app_names = apps
+    # Let's validate we have the needed options defined.
+    ['keyname', 'login', 'table'].each{ |key|
+      unless @options[key]
+        msg = "Error: cannot find #{key} in options!" unless @options[key]
+        Djinn.log_error(msg)
+        return msg
+      end
+    }
 
-    nodes = Djinn.convert_location_array_to_class(locations, keyname)
-    converted_nodes = convert_fqdns_to_ips(nodes)
     @state_change_lock.synchronize {
-      @nodes = converted_nodes
+      @nodes = Djinn.convert_location_array_to_class(locations,
+        @options['keyname'])
     }
     @options = sanitize_credentials()
+    # Let's validate the layout has the needed roles. The following
+    # functions would crash if the role is not there.
+    get_db_master
```

Comment:
I haven't looked closely at all the places where those functions are used. Maybe it's appropriate to crash in some other places that are calling them. For this particular function, though, the pattern seems to be to return an error message if there's a problem with the input that the client provided. I think it would be inconsistent for this function to return an error in some cases and crash in others.